Add tool that shows spectrum at spaxel in cubeviz

javerbukh opened the pull request:
<!-- This comments are hidden when you submit the pull request,
so you do not need to remove them! -->

<!-- Please be sure to check out our code of conduct,
https://github.com/spacetelescope/jdaviz/blob/main/CODE_OF_CONDUCT.md . -->

### Description
<!-- Provide a general description of what your pull request does.
Complete the following sentence and add relevant details as you see fit. -->

<!-- In addition please ensure that the pull request title is descriptive
and allows maintainers to infer the applicable viz component(s). -->

This pull request adds a tool to cubeviz that allows a user to see the spectrum at a spaxel location on the image viewer. The tool is currently housed in the subset section. A spectrum is made for each active layer in the image viewer that is selected (as long as that layer is also a cube).

There are a few questions that need to be answered or things to be fixed, which can be found in the todo section.


https://user-images.githubusercontent.com/7179333/168294410-2d8fb798-a55f-4fb9-9bf7-6c465fabe1b4.mov



<!-- If the pull request closes any open issues you can add this.
If you replace <Issue Number> with a number, GitHub will automatically link it.
If this pull request is unrelated to any issues, please remove
the following line. -->


### Todo
- [ ] Consult @jenneh about spectrum color
- [ ] Consult @jenneh about what icon to use and where to locate it
- [ ] Remove warning `UserWarning: Given trait value dtype "float64" does not match required type "float64". A coerced copy has been created.`
- [ ] Is there a good way to let the user know which pixel was selected or can PR #1315 be used for that instead?
- [ ] Is it possible to link the spectrum to the original cube using wcs? Currently this is done manually using LinkSame
- [ ] The spectrum is currently added using `app.add_data(spec, label)` followed by `app.add_data_to_viewer("spectrum-viewer", label, clear_other_data=False)` but the spectrum does not actually appear visible in the viewer. The current work around is to manually set `figure.marks[-1]` to be visible and correctly scaled, but there should be a better way.

### Checklist for package maintainer(s)
<!-- This section is to be filled by package maintainer(s) who will
review this pull request. -->

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [ ] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [ ] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [ ] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Did the CI pass? If not, are the failures related?
- [ ] Is a change log needed? If yes, is it added to `CHANGES.rst`?
- [ ] Is a milestone set?
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

pllim commented on jdaviz/configs/cubeviz/plugins/tools.py:
```diff
@@ -42,3 +48,97 @@ def on_mouse_event(self, data):
+
+                spec = Spectrum1D(flux=cube.flux[x][y], spectral_axis=cube.spectral_axis)
+                label = f"{data.label}_at_pixel"
+                self.viewer.jdaviz_app.remove_data_from_viewer("spectrum-viewer", label)
+                # self.viewer.session.data_collection[label] = spec
+                self.viewer.jdaviz_app.add_data(spec, label)
+                self.viewer.jdaviz_app.add_data_to_viewer("spectrum-viewer", label, clear_other_data=False)
```
I wonder if glue Data has API to just update the underlying data array rather than completely remove and re-add. 🤔

javerbukh replied:
I'll look into it, that would definitely be helpful!

javerbukh commented:
I hate to be the dark mode police but...the file names in the data drop down are not easily visible.
<img width="453" alt="Screen Shot 2022-05-16 at 2 47 10 PM" src="https://user-images.githubusercontent.com/7179333/168661677-df10a530-0ee2-4936-bd96-1338b46c37f3.png">
.

javerbukh commented:
> Should showing the magenta per-pixel spectrum hide all other spectral entries? Sometimes they overlap and you just see vertical lines from the collapsed full-cube spectrum, which can be confusing.

This is a good point and something I wanted to do originally, but the POs asked for the functionality to mirror a subset. Since a subset can be visible with its reference data, I do the same here. It makes more sense when the collapse function in the spectrum viewer is something other than sum.

> If deactivating the tool resets the zoom on the spectrum viewer, should it also remove the magenta spectrum? Or is the user responsible for finding the entry in the data dropdown and removing it manually. We need some way to "reset", but by doing so when deactivating the tool, it's impossible to use in conjunction with zoom (to zoom in to see what pixel you selected, for example). Instead, you'd have to make sure to zoom first, and then select the pixel.

I guess it depends if a user would want to do continued analysis on the spectrum after it has been created. If yes, then it should remain even after the tool has been deactivated. As for using in conjunction with the zoom tool, you bring up a good point. I'll have to think about that.

> Depending on the answer to the above question, should the created pixel entry be hidden from the user (in https://github.com/spacetelescope/jdaviz/pull/1313)? Otherwise they can toggle it and it loses memory of its color. Should it be available in data dropdowns for input to plugins (like model fitting)?

That is a good question, I think I would have to defer to the POs on that one.

> What should happen when using the tool on a static image (moment map)? Right now it seems to show the highlight on the image but not draw on spectrum. Should we somehow de-activate the tool entirely if no layers in that viewer have a spectral axis?

The highlighted pixel should not appear in that case, I'll fix that in the next commit.

> Should we allow the tool to be used on non-flux cubes (ivar, mask)? If so, do we need to change the "flux"-units on the spectrum-viewer?

The tool already can be used on those cubes. I don't know how much unit conversion I want to get into in this PR but that is a good point. I'll have to think about that and get some feedback from the POs.

javerbukh commented on jdaviz/configs/cubeviz/plugins/viewers.py:
```diff
@@ -20,15 +20,15 @@ class CubevizImageView(BqplotImageView, JdavizViewerMixin):
                     ['bqplot:home'],
                     ['jdaviz:boxzoom', 'bqplot:panzoom'],
-                    ['bqplot:circle', 'bqplot:rectangle'],
+                    ['bqplot:circle', 'bqplot:rectangle', 'jdaviz:spectrumperspaxel'],
```
The toolbar collapsing is not great but I do agree it doesn't exactly fit in the subset group. The main reason it is there is that the original acceptance criteria was that this tool acts like a subset.

kecnry replied:
It acts like a subset, but it doesn't actually create a subset entry.  And it has automatic zooming capability the subsets don't.  If you/Jenn agree it fits better in the new category, I think we can make the toolbar fit by tightening up some of the paddings (it seems we just need a few more pixels).

javerbukh replied:
Ok I moved it into it's own spot.

javerbukh commented on jdaviz/configs/cubeviz/plugins/tools.py:
```diff
@@ -42,3 +46,125 @@ def on_mouse_event(self, data):
+                dc[label].meta["reference_data"] = data.label
+                dc[label].meta["created_from_pixel"] = f"({x}, {y})"
+
+                # Set color of spectrum to magenta to not conflict with subset colors
+                self.spectrum_viewer.figure.marks[-1].colors = self.colors
```
Good point! I will try that.

javerbukh commented on jdaviz/configs/cubeviz/plugins/tools.py:
```diff
@@ -42,3 +46,125 @@ def on_mouse_event(self, data):
+            return
+        world1 = data1.world_component_ids
+        world2 = data2.world_component_ids
+        if len(world1) == 3 and len(world2) == 1:
+            links = [LinkSame(world1[2], world2[0])]
+        elif len(world1) == 1 and len(world2) == 1:
+            links = [LinkSame(world1[0], world2[0])]
+        else:
+            return
```
If the reference data in the spectrum viewer is a 1D spectrum. This happens in the case that the spectrum from the cube is deselected and the spectrum uses itself or another 1D spectrum for linking.

kecnry commented:
> I guess it depends if a user would want to do continued analysis on the spectrum after it has been created. If yes, then it should remain even after the tool has been deactivated.

Ok.  This will have some consideration as well for #1313 - right now that allows data entries to be deleted only if meta['Plugin'] is populated.  We can either edit that logic to detect some of the meta-data you set here, or maybe populate meta['Plugin'] anyways (it sort of is a plugin, it just isn't in the tray).  What do you think? 

> The tool already can be used on those cubes.

Then this is different than the current subset behavior which only collapses subsets on the flux cube.  We could only add the tool to the left-most viewer, but we currently have no logic that ensures that viewer has the flux cube (and that no other viewers do).  Maybe this should be left as-is for now and having data-aware availability of tools should be a follow-up effort (there are other similar situations: the line selection tool in the spectrum viewer should only be available if lines are plotted, etc)?

javerbukh commented:
Yeah I can set that meta data, good idea. Maybe we can check if the cube has flux units? I'm not sure if that is the case for the IVAR cube but I can check that as well.

kecnry reviewed (approved):
Not sure either of these should block this PR, so I'm approving, but might be things we want to consider for follow-up efforts and as the new tool starts to get used in practice.

1. I am torn whether the magenta box on the image viewer should stay persistent when disabling the tool.  Maybe it should be tied to the spectrum data entry (so the only way to get rid of the box would be to _delete_ that data entry)... but I can also see how that might get annoying and in the way and require an extra click.  This might just take us some using this tool in practice and then revise the behavior.

2. The magenta box seems slightly offset from the image itself, but I'm unsure if its the image displaying, box displaying, or higher upstream that is to blame.  See screenshot below.  This is really only apparent if you zoom significantly and its still clear which pixel is highlighted, so I don't think its a major concern.

<img width="242" alt="Screen Shot 2022-05-20 at 9 11 45 AM" src="https://user-images.githubusercontent.com/877591/169535686-40c821ba-6a1a-4ca2-9c4e-62b29aaa9e15.png">

pllim commented:
Re: Magenta box -- Why not make things simpler and draw a "X" at the center of the spaxel ?

javerbukh commented:
@pllim I already had code for a box 😄 if the team prefers an "X" I can change it.

pllim commented:
Slightly offcenter X is less obvious than a misaligned box. Just sayin'. 😉

javerbukh commented:
I switched to the X mark since @pllim makes a good point. @kecnry sorry but I accidentally did some github boneheadedness and removed your commit. Can you please push it one more time?

javerbukh commented:
@kecnry Thank you 😅

PatrickOgle reviewed (changes requested):
Please address the following UI/UX issues before merge, as they can cause confusion.

(1) This  tool works best when the spectral viewer collapse method is 'maximum',
but not when the method is 'sum', because the sum spectrum is off-scale compared to the single spaxel. Need to either 1) revert to maximum, not sum as default  2) change collapse method to maximum when this tool is invoked.

(2) When the tool is deselected, the spaxel dataset should be unchecked but remain in the spectrum data dropdown. That way, it can be operated on, but it is clear that we are no longer in spaxel viewing  mode.

(3) Can we allow two tools to be selected, e.g. spaxel tool and magnify, spaxel tool and region selection, spaxel tool and plot options? This should only be for spaxel+one other tool, because some other combinations have conflicting cursor operation.

(4) In a similar vein, there is strange behavior when selecting plot options if spaxel tool is selected.  It opens the Plot Plugin, then flashes back to highlighting the spaxel tool.  This would be fixed if (3) above is implemented.

PatrickOgle commented:
Should not rescale spectrum viewer on each pixel click.  This destroys the context of whether the selected pixel is brighter or fainter than the previously selected one and the reference spectrum.

javerbukh commented:
> (1) This tool works best when the spectral viewer collapse method is 'maximum', but not when the method is 'sum', because the sum spectrum is off-scale compared to the single spaxel. Need to either 1) revert to maximum, not sum as default 2) change collapse method to maximum when this tool is invoked.

Changing the function depending on the tool does not feel intuitive. Reverting the default function can be a separate ticket since it sounds like the POs need to have a discussion about that.

> (2) When the tool is deselected, the spaxel dataset should be unchecked but remain in the spectrum data dropdown. That way, it can be operated on, but it is clear that we are no longer in spaxel viewing mode.

My only counter to this is that plugins create data that is made visible in the viewer and then can only be deselected by the user doing into the data dropdown and doing it manually. For example, this is how the model fitting plugin works.

> (3) Can we allow two tools to be selected, e.g. spaxel tool and magnify, spaxel tool and region selection, spaxel tool and plot options? This should only be for spaxel+one other tool, because some other combinations have conflicting cursor operation.

As far as I know this is not yet implemented. @kecnry do you know how difficult that would be to implement?

> (4) In a similar vein, there is strange behavior when selecting plot options if spaxel tool is selected. It opens the Plot Plugin, then flashes back to highlighting the spaxel tool. This would be fixed if (3) above is implemented.

I think this is intended behavior. The same interaction happens with all other tools.

> (5)  Should not rescale spectrum viewer on each pixel click. This destroys the context of whether the selected pixel is brighter or fainter than the previously selected one and the reference spectrum.

If a user has their collapse function set to `Sum`, the spectrum can look like a flat line at `0` and they will assume it's not working. Or, a user could have the function set to `Minimum` or `Mean` and not see the spectrum above the reference spectrum. I can make a separate PR to test that functionality and we can decide after testing it.

There is no way to compare the current pixel's spectrum to a previous one. When another pixel is clicked, the spectrum is replaced by the spectrum at the new pixel. I could see a shift-click paradigm for highlighting multiple pixels and showing the spectrum at each point, but I think that would be an enhancement to this feature.

I will be out starting next Tuesday. If some of these suggestions can be broken out in to other tickets so they can be discussed and planned that would be great. It sounds like `Tool multiselect`, `Revert collapse function to max`, and `Shift-click to see multiple spectra` are some that you mention. I can open a new PR to show the functionality you suggested in (2) and (5), namely, deselecting the spectrum after deactivating the tool and not adjusting the viewer limits when the spectrum is created.

pllim commented:
Some random thoughts...

> Can we allow two tools to be selected

Instead of selecting multiple tools, we can also tie this spaxel selection to a key binding, not unlike Imviz line profile. That is a completely separate operation from tools and is "always on" (can be good or bad).

> compare the current pixel's spectrum to a previous one

Maybe it is good enough to have an API to allow you to extract the pixel's spectrum at the desired location out as `Spectrum1D` back into the notebook and then you can compare them outside of Jdaviz to your heart's content?

kecnry commented:
I'll just comment about "tool multiselect" - that is intentionally not implemented.  I suppose we could have one tool handle drag events and the other click events, but the distinction between those two is not always clear and I think it would be confusing for the user (and for the underlying infrastructure).  

The only idea I have if the "x" is tied to the created spectrum data-entry, and so shows up in the data-dropdown of the image viewer as well and is rendered as an "x" at that pixel because of the coordinates in the meta-data.  That way, clicking on another tool (like plot options or zoom) and then returning to the pixel select, would not lose the marker.  I would say then that the visibility of that dataset should either turn off in both the image viewer and the spectrum viewer when the tool is inactive, or not for both (and leave it up to the user to uncheck one or both).

PatrickOgle reviewed (approved):
Not changing the plot scaling every time a new spaxel is selected helped a lot.  Now there is a fixed reference frame for understanding differences in spaxel flux.

The spaxel spectrum plotting is a bit laggy and can cause confusion for the first spaxel selection.  "Where is my spaxel spectrum?"

My remaining suggestions (primarily allowing two tools to be active) can be addressed in separate tickets.

kecnry commented on jdaviz/configs/cubeviz/plugins/tools.py:
```diff
@@ -42,3 +47,133 @@ def on_mouse_event(self, data):
+                # Remove data from viewer, re-add data to app, and then add data
+                # back to spectrum viewer
+                self.viewer.jdaviz_app.remove_data_from_viewer("spectrum-viewer", label)
+                dc[label] = spec
```
is there any difference between this and calling `app.add_data` (looks like the only thing calling `add_data` would add is some label checks and a snackbar message)?

duytnguyendtn commented on jdaviz/configs/cubeviz/plugins/tools.py:
```diff
@@ -42,3 +47,133 @@ def on_mouse_event(self, data):
+            x = round(event_x)
+            y = round(event_y)
+
+            self.remove_highlight_on_pixel()
+            self.highlight_pixel(x, y)
```
Whats the reasoning behind needing to remove the highlight on the pixel before readding it?

javerbukh replied:
If you do not do that, you see multiple pixels with the "X" over it at every location where the use clicked, which is confusing.

duytnguyendtn commented on jdaviz/configs/cubeviz/plugins/tools.py:
```diff
@@ -42,3 +47,133 @@ def on_mouse_event(self, data):
+                                  meta=cube.meta)
+                label = f"{data.label}_at_pixel"
+
+                # Remove data from viewer, re-add data to app, and then add data
+                # back to spectrum viewer
+                self.viewer.jdaviz_app.remove_data_from_viewer("spectrum-viewer", label)
+                dc[label] = spec
+                self.viewer.jdaviz_app.add_data_to_viewer("spectrum-viewer",
+                                                          label,
+                                                          clear_other_data=False)
+                self.created_data_labels.append(label)
```
I assume this section is to wipe the old spectrum and replace it with the new spectrum? It wasn't clear from the comment. Because there are other places in past where we've had to "toggle" the same thing on and off to get things to work, it might be worth saying that this is NOT one of those scenarios

javerbukh replied:
I don't think I understand. What would you like changed?

duytnguyendtn commented on jdaviz/configs/cubeviz/plugins/tools.py:
```diff
@@ -42,3 +47,133 @@ def on_mouse_event(self, data):
+            self.viewer.jdaviz_app.remove_data_from_viewer("spectrum-viewer", label)
+        self.created_data_labels = []
+
+    def on_mouse_event(self, data):
+        event = data['event']
+
+        if event == 'click':
```
I've never had to watch for mouse events before, but this feels heavy handed (this method is going to fire for all mouse events then check to see if it's a click event). Is there a way to monitor for ONLY click events? Trying to think about performance

pllim replied:
Yes, that is basically what it is doing under `activate` method. The `if` is still good to have in case we want to expand this logic to other mouse/key events.

```python
        self.viewer.add_event_callback(self.on_mouse_event,
                                       events=['click'])
```

kecnry reviewed (commented):
Re-reviewing since there were significant changes since my original review.  Sorry if some of these comments were in code that existed then too!

duytnguyendtn reviewed (commented):
I share the same concerns the other devs have pointed out above, but also agree this is in a shippable state where we can begin testing with our users! Thanks for the feature @javerbukh! A few concerns:

1) The icon is a little small in the plugin bar to the point it's hard to read. I assume @Jenneh designed it, in which case I'm willing to trust her here, but it is a little difficult for me to discern what the icon actually looks like
2) Since we can click on multiple image windows, I was able to get into a state where I saw two spectrums for each spaxel I clicked in different windows. I couldn't find a way to discern the two of them since they're both magenta and we strip those options from the plot options. This seems like a big enough UX concern that we might want to consider it before we merge?
![image](https://user-images.githubusercontent.com/25206008/170752702-62e6a9d6-2ed3-40bc-8299-172efec85f8a.png)

javerbukh commented on jdaviz/configs/cubeviz/plugins/tools.py:
```diff
@@ -42,3 +47,133 @@ def on_mouse_event(self, data):
+                dc[label].meta["Plugin"] = "Spectrum per spaxel"
+
+                # Set color of spectrum to not conflict with subset colors
+                self.spectrum_viewer.state.layers[-1].color = self.colors[0]
```
That's one possibility. Also if we want to use a color other than magenta we do not have to change it in two places.

javerbukh commented:
I tried to address all review comments that I could. If something was left out that you feel strongly about, please let me know. Otherwise this should be ready for final reviews.

pllim commented on jdaviz/configs/cubeviz/plugins/tools.py:
```diff
@@ -3,10 +3,16 @@
+from jdaviz.core.marks import SelectedPixel
+from jdaviz.core.events import SnackbarMessage
+
+
+__all__ = ['SelectSlice', 'SpectrumPerSpaxel']
```
I tried to rebase this PR locally so I can review it on top of latest changes and I noticed this. I purposely make this empty list in #1346 to have this be consistent with other configs.

```suggestion
__all__ = []
```

javerbukh replied:
Got it, I'll commit it then.

pllim commented on docs/cubeviz/displaycubes.rst:
```diff
@@ -110,6 +110,19 @@ of the selected area, but will expand or limit the y range to preserve the aspec
+
+This tool allows a user to select one pixel in the image viewer and visualize the spectrum
+at that point in the spectrum viewer. The selected pixel will be highlighted in the image viewer
+and the spectrum will have the same color. If there are multiple data selected in the image viewer
```
I don't think it makes sense to even allow this on clicks outside of flux-viewer, since the spectrum-viewer only corresponds to flux-viewer. It gets too confusing to see spaxel from, say, uncert-viewer shown but we don't have a way to collapse anything outside of flux-viewer.

javerbukh replied:
I'll leave this up to the POs to decide as a follow-up ticket.

pllim commented on jdaviz/configs/cubeviz/plugins/tests/test_tools.py:
```diff
@@ -0,0 +1,60 @@
+@pytest.mark.filterwarnings('ignore:No observer defined on WCS')
+def test_spectrum_at_spaxel(cubeviz_helper, spectrum1d_cube):
+    app = cubeviz_helper.app
+    app.add_data(spectrum1d_cube, 'test')
+    app.add_data_to_viewer("spectrum-viewer", "test")
+    app.add_data_to_viewer("flux-viewer", "test")
```
Should we do this to reflect how the API is actually used?
```suggestion
    cubeviz_helper.load_data(spectrum1d_cube, data_label='test')
```

pllim commented on jdaviz/configs/cubeviz/plugins/tools.py:
```diff
@@ -42,3 +47,146 @@ def on_mouse_event(self, data):
+    def activate(self):
+        self.viewer.add_event_callback(self.on_mouse_event,
+                                       events=['click'])
+        self.spectrum_viewer = self.viewer.jdaviz_app.get_viewer_by_id("cubeviz-3")
```
The tests suggests that you grab by reference. Why impose ID here and only here?
```suggestion
        self.spectrum_viewer = self.viewer.jdaviz_app.get_viewer("spectrum-viewer")
```

pllim commented on jdaviz/configs/cubeviz/plugins/tools.py:
```diff
@@ -42,3 +47,146 @@ def on_mouse_event(self, data):
+        event = data['event']
+
+        if event == 'click':
+            event_x = data['domain']['x']
+            event_y = data['domain']['y']
```
If you ever wanna support multiple cubes in the same viewer (gasp), keep in mind that this is the X/Y of the *reference data*. FYI.

pllim commented on jdaviz/configs/cubeviz/plugins/tools.py:
```diff
@@ -42,3 +47,146 @@ def on_mouse_event(self, data):
+                cube_in_viewer = True
+
+                cube = data.get_object(cls=Spectrum1D, statistic=None)
+                if x > cube.shape[0] or y > cube.shape[1] or x < 0 or y < 0:
```
```suggestion
                if x >= cube.shape[0] or y >= cube.shape[1] or x < 0 or y < 0:
```

pllim commented on jdaviz/configs/cubeviz/plugins/tools.py:
```diff
@@ -42,3 +47,133 @@ def on_mouse_event(self, data):
+                # Remove data from viewer, re-add data to app, and then add data
+                # back to spectrum viewer
+                self.viewer.jdaviz_app.remove_data_from_viewer("spectrum-viewer", label)
+                dc[label] = spec
```
Tsk tsk, maybe we need a `app.add_data(..., shhhh=True)`.

pllim commented on jdaviz/configs/cubeviz/plugins/tools.py:
```diff
@@ -42,3 +47,146 @@ def on_mouse_event(self, data):
+                                  meta=cube.meta)
+                # Add meta data for reference data and pixel
+                spec.meta["reference_data"] = data.label
+                spec.meta["created_from_pixel"] = f"({x}, {y})"
```
To be consistent?
```suggestion
                spec.meta["created_from_spaxel"] = f"({x}, {y})"
```

pllim commented on jdaviz/configs/cubeviz/plugins/tools.py:
```diff
@@ -42,3 +47,146 @@ def on_mouse_event(self, data):
+                spec.meta["created_from_pixel"] = f"({x}, {y})"
+                spec.meta["Plugin"] = "Spectrum per spaxel"
+
+                label = f"{data.label}_at_pixel"
```
To be consistent?
```suggestion
                label = f"{data.label}_at_spaxel"
```

pllim commented on jdaviz/configs/cubeviz/plugins/tools.py:
```diff
@@ -42,3 +47,146 @@ def on_mouse_event(self, data):
+            # Flux viewer is empty or contains only 2D images
+            if not cube_in_viewer:
+                self.remove_highlight_on_pixel()
+                msg = SnackbarMessage("No cube in viewer, cannot create spectrum from pixel",
```
To be consistent.
```suggestion
                msg = SnackbarMessage("No cube in viewer, cannot create spectrum from spaxel",
```

pllim commented on jdaviz/configs/cubeviz/plugins/tools.py:
```diff
@@ -42,3 +47,146 @@ def on_mouse_event(self, data):
+
+        self.viewer.session.data_collection.add_link(links)
+
+    def highlight_pixel(self, x, y):
```
So, is it pixel or spaxel? Maybe we should pick one and stick with it?

pllim commented on jdaviz/configs/cubeviz/plugins/tools.py:
```diff
@@ -42,3 +47,146 @@ def on_mouse_event(self, data):
+
+    def highlight_pixel(self, x, y):
+        # Creates an X that shows the pixel at coordinate (x, y)
+        x_coords = [[x - 0.5, x + 0.5], [x - 0.5, x + 0.5]]
+        y_coords = [[y - 0.5, y + 0.5], [y + 0.5, y - 0.5]]
```
I see repeated math. Should we only calculate each unique equation once and reuse?

pllim commented on jdaviz/configs/cubeviz/plugins/tools.py:
```diff
@@ -42,3 +47,146 @@ def on_mouse_event(self, data):
+        self.viewer.figure.marks = self.viewer.figure.marks + [highlight]
+
+    def update_highlight_pixel(self, x, y):
+        x_coords = [[x - 0.5, x + 0.5], [x - 0.5, x + 0.5]]
+        y_coords = [[y - 0.5, y + 0.5], [y + 0.5, y - 0.5]]
```
Again, repeated here.
```suggestion
        x_coords = [[x - 0.5, x + 0.5], [x - 0.5, x + 0.5]]
        y_coords = [[y - 0.5, y + 0.5], [y + 0.5, y - 0.5]]
```

pllim commented on jdaviz/core/marks.py:
```diff
@@ -468,3 +468,8 @@ def __init__(self, **kwargs):
         super().__init__(**kwargs)
+
+
+class SelectedPixel(Lines):
```
Please add this to the `__all__` listing above.

pllim reviewed (commented):
Should we add a note in the user doc that if you see spaxel spectrum as flat, try zoom in a lot because the signal/flux not summed like the original spectrum?

From the code, I expected the magenta stuff to auto-clear when I collapse the cube and only display the collapsed image in flux viewer, but it didn't. Did I misunderstand? Here is what I see.

![Screenshot 2022-06-01 144318](https://user-images.githubusercontent.com/2090236/171479288-f7c70327-15ff-4db5-a71a-1a827ac7e7c9.png)
![Screenshot 2022-06-01 144334](https://user-images.githubusercontent.com/2090236/171479289-2c376446-539a-479c-8e0f-7d5d9d2ffaaf.png)

pllim commented on jdaviz/configs/cubeviz/plugins/tools.py:
```diff
@@ -42,3 +47,150 @@ def on_mouse_event(self, data):
+                                    if not isinstance(x, SelectedSpaxel)]
+
+    def _calc_coords(self, x, y):
+        x_coords = [[x - 0.5, x + 0.5], [x - 0.5, x + 0.5]]
+        y_coords = [[y - 0.5, y + 0.5], [y + 0.5, y - 0.5]]
```
Pure nitpick. I'll leave it up to you to accept or ignore.
```suggestion
        xmin = x - 0.5
        xmax = x + 0.5
        ymin = y - 0.5
        ymax = y + 0.5
        x_coords = [[xmin, xmax], [xmin, xmax]]
        y_coords = [[ymin, ymax], [ymax, ymin]]
```

pllim reviewed (approved):
I think this is a very good first pass, and we should get this in sooner than later. Thanks for your patience!
